### ISSUE TITLE
Inferenced images do not match samples during training

### ISSUE THREAD

**Issue by Ianmcmill** [akanimax/pro_gan_pytorch#16]
Inferenced images generated with this code
```
import torch as th
import pro_gan_pytorch.PRO_GAN as pg
import torchvision.utils as vutils

fdepth = 7
depth = 4
latent_size = 512

device = th.device("cuda" if th.cuda.is_available() 
                   else "cpu")
gen = th.nn.DataParallel(pg.Generator(fdepth, latent_size))
gen.load_state_dict(th.load("training_runs/portrait_large_homogen_256/saved_models/GAN_GEN_4.pth", map_location=str(device)))
for x in range(5): 
    noise = th.randn(1, 512).to(device)
    sample_image = gen(noise, depth, alpha=1).detach()
    vutils.save_image(sample_image[0, :, :, :], 'portaits_' + str(x) + '_' + str(depth) + '_' + str(latent_size) + '.png'.format(4))
```
Result (arranged and upsampled with Photoshop)
![inference_collage](https://user-images.githubusercontent.com/10505658/52403116-43268580-2ac6-11e9-9824-f2ad505d8c5e.png)

Do not resemble generated samples during training:
![gen_4_245_48](https://user-images.githubusercontent.com/10505658/52403145-56395580-2ac6-11e9-9a6a-3f811d7c50bb.png)

What is wrong with the code?
The inference samples do not resemble in color nor in shape.
Should I avoid saving images with torchvision???
I can provide the .pth model if desired.

**Comment by Ianmcmill** [comment]
```
for x in range(1): 
    noise = th.randn(1, 512).to(device)
    sample_image = gen(noise, depth, alpha=1).detach()
    #sample_grid = vutils.make_grid(sample_image, nrow=8, padding=2, )
    vutils.save_image(sample_image[:, :, :, :], 'portaits_' + str(x) + '_1_1_' + str(depth) + '_' + str(latent_size) + '.png'.format(3), normalize=True, padding=0)
```

Also `device = th.device("cuda" if th.cuda.is_available() else "cpu")`. If no CUDA is available CPU does not work because the network structure is not compatible. This line can be changed to `device = th.device("cuda")` as the supplied training code lacks CPU implementation. So sampling with CPU is not implemented.

**Comment by akanimax** [comment]
Hi @Ianmcmill,

Apologies for late reply. I was travelling for last two weeks. 
You are not normalizing the generated images in the [0 - 1 ] range in order for the display packages to show them correctly.
Please use this technique to normalize them: (assuming img is [1024 x 1024 x 3] array):

    img = (img - img.min()) / (img.max() - img.min())

Hope this helps

Best regards,
@akanimax

**Comment by Ianmcmill** [comment]
I have no idea what you are talking about :sweat_smile:
Nevermind I'll try something else. Thanks for you help.

**Comment by maxvfischer** [comment]
I'm hijacking this issue.

I've trained my model on abstract art, but similar to @Ianmcmill, the generated images doesn't resemble the samples generated during training.

I've normalized them per your recommendation, but they still differs from the sampled images.

```
Config used during training of ConditionalProGAN:
    depth = 9
    num_epochs = [27, 54, 54, 54, 54, 54, 54, 54, 54]
    fade_ins = [50, 50, 50, 50, 50, 50, 50, 50, 50]
    batch_sizes = [16, 16, 16, 16, 16, 16, 14, 6, 3]
    latent_size = 512
```

The training is currently training on the last resolution, 1024x1024. 

I'm using the following code to generate an image:

```
import torch as th
import pro_gan_pytorch.PRO_GAN as pg
import matplotlib.pyplot as plt
from collections import OrderedDict
import os

os.environ['KMP_DUPLICATE_LIB_OK']='True'
fdepth = 9
depth = 8

device = th.device("cuda" if th.cuda.is_available()
                   else "cpu")
gen = pg.Generator(depth=fdepth, latent_size=512,
                   use_eql=False).to(device)

state_dict = th.load("/path/progan/GAN_GEN_SHADOW_8.pth", map_location='cpu')

new_state_dict = OrderedDict()
for k, v in state_dict.items():
    name = k[7:] # remove `module.`
    new_state_dict[name] = v

gen.load_state_dict(
    new_state_dict
)

noise = th.randn(1, 512).to(device)

sample_image = gen(noise, depth=8, alpha=1).detach()

sample_image = (sample_image - sample_image.min()) / (sample_image.max() - sample_image.min())
plt.imshow(sample_image[0].permute(1, 2, 0) / 2 + 0.5)
plt.show()
```

"GAN_GEN_SHADOW_8.pth" is the last step currently trained. 

What might be the problem?

**Sampled images during training**
![gen_8_13_1632](https://user-images.githubusercontent.com/28877367/59159787-3abf7500-8acf-11e9-93ef-f7efb8f325ec.png)

**Generated images with trained weights**
<img width="370" alt="Screenshot 2019-06-09 at 15 54 52" src="https://user-images.githubusercontent.com/28877367/59159792-4dd24500-8acf-11e9-861d-4868398fae32.png">
<img width="369" alt="Screenshot 2019-06-09 at 15 55 07" src="https://user-images.githubusercontent.com/28877367/59159796-4f9c0880-8acf-11e9-8642-34234a1cd030.png">
<img width="370" alt="Screenshot 2019-06-09 at 15 55 21" src="https://user-images.githubusercontent.com/28877367/59159797-532f8f80-8acf-11e9-86c0-54d74c5e8ba4.png">
